src/devices/src/virtio/block/device.rs

```diff
@@ -332,886 +332,574 @@ impl VirtioDevice for Block {
-//                 );
-//             }
-//         }
-
-//         // test block device update handler
-//         {
-//             let f = NamedTempFile::new().unwrap();
-//             let path = f.path().to_path_buf();
-//             let mdata = metadata(&path).unwrap();
-//             let mut id = vec![0; VIRTIO_BLK_ID_BYTES as usize];
-//             let str_id = format!("{}{}{}", mdata.st_dev(), mdata.st_rdev(), mdata.st_ino());
-//             let part_id = str_id.as_bytes();
-//             id[..cmp::min(part_id.len(), VIRTIO_BLK_ID_BYTES as usize)].clone_from_slice(
-//                 &part_id[..cmp::min(part_id.len(), VIRTIO_BLK_ID_BYTES as usize)],
-//             );
-
-//             let file = OpenOptions::new()
-//                 .read(true)
-//                 .write(true)
-//                 .open(path)
-//                 .unwrap();
-//             h.update_disk_image(file).unwrap();
-
-//             assert_eq!(h.disk_image.metadata().unwrap().st_ino(), mdata.st_ino());
-//             assert_eq!(h.disk_image_id, id);
-//         }
-//     }
-// }
+#[cfg(test)]
+pub(crate) mod tests {
+    extern crate vmm_sys_util;
+    use self::vmm_sys_util::tempfile::TempFile;
+
+    use super::*;
+    use crate::virtio::queue::tests::*;
+    use memory_model::GuestAddress;
+    use std::fs::{metadata, OpenOptions};
+    use std::thread;
+    use std::time::Duration;
+    use std::u32;
+
+    /// Will read $metric, run the code in $block, then assert metric has increased by $delta.
+    macro_rules! check_metric_after_block {
+        ($metric:expr, $delta:expr, $block:expr) => {{
+            let before = $metric.count();
+            let _ = $block;
+            assert_eq!($metric.count(), before + $delta, "unexpected metric value");
+        }};
+    }
+
+    impl Block {
+        fn set_queue(&mut self, idx: usize, q: Queue) {
+            self.queues[idx] = q;
+        }
+
+        fn set_rate_limiter(&mut self, rl: RateLimiter) {
+            self.rate_limiter = rl;
+        }
+
+        fn rate_limiter(&mut self) -> &RateLimiter {
+            &self.rate_limiter
+        }
+    }
+
+    /// Create a default Block instance to be used in tests.
+    pub(crate) fn default_block() -> Block {
+        // Create backing file.
+        let f = TempFile::new().unwrap();
+        f.as_file().set_len(0x1000).unwrap();
+
+        // Rate limiting is enabled but with a high operation rate (10 million ops/s).
+        let rate_limiter = RateLimiter::new(0, None, 0, 100_000, None, 10).unwrap();
+
+        let mem = GuestMemory::new(&[(GuestAddress(0), 0x10000)]).unwrap();
+
+        Block::new(
+            mem.clone(),
+            f.as_file().try_clone().unwrap(),
+            true,
+            rate_limiter,
+        )
+        .unwrap()
+    }
+
+    fn invoke_handler_for_queue_event(b: &mut Block) {
+        // Leave at least one event here so that reading it later won't block.
+        b.interrupt_evt.write(1).unwrap();
```

Comment:
Why do we have to leave at least one event here so that reading it later won't block?
Shouldn't the signal be handled by the `process_queue_event`, and reading it later not be a problem?
This would mean that the later `assert_eq!`, on the `interrupt_evt.read()`, should be tested against `1`, not `2`.
The same happens in net device `test_handler` unit test. Is this intentional, or could we change this to be cleaner?

Reply:
I think the reasoning behind this is if the test fails (we don't get to write 1 to `interupt_evt` as expected), then trying to read from an empty fd would panic.

Reply:
This was actually required before when our `EventFd`s were blocking.
It is no longer required if the `EventFd` was opened with the `libc::EFD_NONBLOCK` flag.

Reading on an empty non-blocking `EventFd` will just return `std::io:Err(EAGAIN)` instead of blocking.

Reply:
Removed this.